File: apps/mobile/lib/models/registration.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import '../core/firestore_converters.dart';
import 'enums.dart';

/// Identité du joueur soumise via l'app register (`/registrations.player`).
class RegistrationPlayerIdentity {
  const RegistrationPlayerIdentity({
    required this.firstName,
    required this.lastName,
    required this.birthDate,
    required this.gender,
    required this.avs,
    required this.avsUnavailable,
    required this.phone,
  });

  final String firstName;
  final String lastName;
  final DateTime? birthDate;

  /// "M" | "F" | "other" | null — gardé en string brut (peu d'usage mobile).
  final String? gender;
  final String? avs;
  final bool avsUnavailable;
  final String? phone;

  String get fullName => '$firstName $lastName'.trim();

  factory RegistrationPlayerIdentity.fromMap(Map<String, dynamic> map) =>
      RegistrationPlayerIdentity(
        firstName: FsConvert.str(map['firstName']),
        lastName: FsConvert.str(map['lastName']),
        birthDate: FsConvert.toDateTime(map['birthDate']),
        gender: FsConvert.strOrNull(map['gender']),
        avs: FsConvert.strOrNull(map['avs']),
        avsUnavailable: FsConvert.boolean(map['avsUnavailable']),
        phone: FsConvert.strOrNull(map['phone']),
      );
}

/// Document `/registrations/{registrationId}` — demande d'inscription
/// self-service. L'app mobile coach lit les inscriptions de ses équipes.
class Registration {
  const Registration({
    required this.id,
    required this.submittedByUid,
    required this.registrationFor,
    required this.player,
    required this.matchedMemberId,
    required this.teamId,
    required this.previouslyLicensed,
    required this.foreignTransfer,
    required this.status,
    required this.statusUpdatedAt,
    required this.trialStartedAt,
    required this.refusalReason,
    required this.refusedByUid,
    required this.createdAt,
  });

  final String id;

  /// uid de l'auteur de l'inscription.
  final String submittedByUid;
  final RegistrationFor registrationFor;
  final RegistrationPlayerIdentity player;

  /// Lien à un `/members/{id}` existant ; `null` = nouveau dossier.
  final String? matchedMemberId;
  final String teamId;
  final bool previouslyLicensed;
  final bool foreignTransfer;
  final RegistrationStatus status;
  final DateTime? statusUpdatedAt;
  final DateTime? trialStartedAt;
  final String? refusalReason;
  final String? refusedByUid;
  final DateTime createdAt;

  factory Registration.fromMap(String id, Map<String, dynamic> map) =>
      Registration(
        id: id,
        submittedByUid: FsConvert.str(map['submittedByUid']),
        registrationFor: RegistrationFor.fromWire(map['registrationFor']),
        player: RegistrationPlayerIdentity.fromMap(
          FsConvert.mapOrNull(map['player']) ?? const {},
        ),
        matchedMemberId: FsConvert.strOrNull(map['matchedMemberId']),
        teamId: FsConvert.str(map['teamId']),
        previouslyLicensed: FsConvert.boolean(map['previouslyLicensed']),
        foreignTransfer: FsConvert.boolean(map['foreignTransfer']),
        status: RegistrationStatus.fromWire(map['status']),
        statusUpdatedAt: FsConvert.toDateTime(map['statusUpdatedAt']),
        trialStartedAt: FsConvert.toDateTime(map['trialStartedAt']),
        refusalReason: FsConvert.strOrNull(map['refusalReason']),
        refusedByUid: FsConvert.strOrNull(map['refusedByUid']),
        createdAt: FsConvert.toDateTimeOr(map['createdAt']),
      );

  factory Registration.fromFirestore(
    DocumentSnapshot<Map<String, dynamic>> snapshot,
  ) =>
      Registration.fromMap(snapshot.id, FsConvert.requireData(snapshot));

  Registration copyWith({
    String? submittedByUid,
    RegistrationFor? registrationFor,
    RegistrationPlayerIdentity? player,
    String? matchedMemberId,
    String? teamId,
    bool? previouslyLicensed,
    bool? foreignTransfer,
    RegistrationStatus? status,
    DateTime? statusUpdatedAt,
    DateTime? trialStartedAt,
    String? refusalReason,
    String? refusedByUid,
    DateTime? createdAt,
  }) =>
      Registration(
        id: id,
        submittedByUid: submittedByUid ?? this.submittedByUid,
        registrationFor: registrationFor ?? this.registrationFor,
        player: player ?? this.player,
        matchedMemberId: matchedMemberId ?? this.matchedMemberId,
        teamId: teamId ?? this.teamId,
        previouslyLicensed: previouslyLicensed ?? this.previouslyLicensed,
        foreignTransfer: foreignTransfer ?? this.foreignTransfer,
        status: status ?? this.status,
        statusUpdatedAt: statusUpdatedAt ?? this.statusUpdatedAt,
        trialStartedAt: trialStartedAt ?? this.trialStartedAt,
        refusalReason: refusalReason ?? this.refusalReason,
        refusedByUid: refusedByUid ?? this.refusedByUid,
        createdAt: createdAt ?? this.createdAt,
      );
}

File: packages/shared-types/src/registration.ts
import type { Timestamp } from './index'

/**
 * Document `/registrations/{registrationId}` — demande d'inscription self-service
 * créée via l'app `apps/courtbase-register`.
 * Voir `docs/chantier-registrations.md` §7.1 pour le détail produit.
 *
 * Les transitions de status post-`submitted` passent par callables (refuseRegistration,
 * onRegistrationStatusChanged, etc.) — pas de write client.
 */

export type RegistrationStatus =
  | 'draft'
  | 'submitted'
  | 'open_pending_trial'
  | 'conditional_pending_review'
  | 'conditional_pending_trial'
  | 'trial_in_progress'
  | 'confirmed_pending_dues'
  | 'active'
  | 'refused'
  | 'cancelled'

export type RegistrationFor = 'self' | 'dependent'

export type RegistrationRelationship =
  | 'parent'
  | 'legal_guardian'
  | 'sibling'
  | 'caritas'
  | 'other'

export type RegistrationActionType =
  | 'created'
  | 'submitted'
  | 'status_changed'
  | 'team_changed'
  | 'refused'
  | 'document_uploaded'

export interface RegistrationActionLogEntry {
  at: Timestamp
  byUid: string
  action: RegistrationActionType
  previousStatus: RegistrationStatus | null
  newStatus: RegistrationStatus | null
  note: string | null
}

/**
 * Identité du joueur soumise via l'app register. Transmise telle quelle au moment
 * de la registration ; un `/members/{id}` n'est créé qu'à l'acceptation par le coach
 * (sauf si `matchedMemberId` lie à un member existant).
 */
export interface RegistrationPlayerIdentity {
  firstName: string
  lastName: string
  birthDate: Timestamp
  gender: 'M' | 'F' | 'other' | null
  /**
   * AVS au format 756.XXXX.XXXX.XX. `null` uniquement au stade `draft` (champ
   * pas encore rempli) — la callable `submitRegistration` rejette toute
   * soumission sans AVS valide : un joueur sans AVS ne peut pas s'inscrire via
   * le portail (cas asile / transfert étranger → traités hors portail).
   */
  avs: string | null
  /** Téléphone direct du joueur (recommandé, pas obligatoire). */
  phone: string | null
}

export interface RegistrationData {
  /** UID de l'auteur de la registration (créateur du compte register). */
  submittedByUid: string
  /** Type d'inscription : pour soi-même (majeur) ou pour un dépendant. */
  registrationFor: RegistrationFor
  /** Lien de parenté ou rôle, `null` si `registrationFor === 'self'`. */
  relationship: RegistrationRelationship | null
  /** Texte libre si `relationship === 'other'`, sinon `null`. */
  relationshipOther: string | null

  player: RegistrationPlayerIdentity

  /**
   * Lien à un `/members/{id}` existant, rattaché par match AVS exact confirmé
   * par l'utilisateur dans le wizard.
   * `null` = nouveau dossier qui sera créé à l'acceptation par le coach.
   */
  matchedMemberId: string | null

  /** ID de l'équipe choisie. */
  teamId: string

  // Historique sportif
  previouslyLicensed: boolean
  previousClubName: string | null
  previousClubAbroad: boolean
  /** Storage path de la lettre de sortie, `null` si pas uploadée. */
  transferLetterStoragePath: string | null
  /** Flag transverse, peut coexister avec n'importe quel `status`. */
  foreignTransfer: boolean

  // Lifecycle
  status: RegistrationStatus
  statusUpdatedAt: Timestamp
  trialStartedAt: Timestamp | null
  refusalReason: string | null
  refusedByUid: string | null

  /** Append-only log des transitions / actions. */
  actionLog: RegistrationActionLogEntry[]

  // Tracking notifs
  coachNotifiedAt: Timestamp | null
  adminNotifiedAt: Timestamp | null

  createdAt: Timestamp
}

export type Registration = RegistrationData & { id: string }

/**
 * Document `/teams/{teamId}/refusalLogs/{logId}` — audit des refus d'inscriptions
 * par les coachs. Écrit uniquement par la callable `refuseRegistration` (Admin SDK).
 * Lecture admin uniquement (audit anti-abus).
 */
export interface RefusalLogData {
  registrationId: string
  /** Dénormalisé pour faciliter le debug / l'audit. */
  playerName: string
  /** Motif libre, obligatoire (validation callable). */
  reason: string
  refusedAt: Timestamp
  /** UID du coach qui refuse. */
  refusedByUid: string
}

export type RefusalLog = RefusalLogData & { id: string }
